docs(js): streamline React Router v6 and v7 documentation (#13138)
getsentry/sentry-docs: docs/platforms/javascript/guides/react/features/react-router/v6.mdx

sidebar_order: 20
---

> 

Reviewer: Would it make sense to lead with something like this:

```suggestion
You can instrument your React Router v6 setup with Sentry. Depending on how you use react-router, slightly different setups are needed:

<TableOfContents ignoreIds={["custom-error-boundary", "next-steps"]} />

Additionally, make sure to [Setup a Custom Error Boundary](#custom-error-boundary) to ensure rendering errors are captured to Sentry.
```

Reviewer: and something similar for react-router v7 :)

Author: @mydea Good point! I updated the start of the page with an adapted version of your suggestion.

Sadly, the TableOfContents component does not work with our structure as it seems to be created to work with the SDK options pages, and our structure here is quite different -> I created an issue #13159  so we can consider upgrading the functionality of this component

Reviewer: should be fixed here https://github.com/getsentry/sentry-docs/pull/13171 😅

Author: the toc is now included - looks good!

> _React Router v6 support is included in the `@sentry/react` package since version `7`._

Reviewer: ```suggestion
```

IMHO we can drop this, we are on v9 now 😅

### Usage with `createBrowserRouter` or `createMemoryRouter`

> To instrument your React Router, update to `Sentry.reactRouterV6BrowserTracingIntegration` within `Sentry.init` and provide the required React hooks and router functions. Then, wrap the router instance created by `createBrowserRouter` or `createMemoryRouter` with one of the following functions:

Reviewer 2: ```suggestion
To instrument your React Router, update your `Sentry.browserTracingIntegration` to `Sentry.reactRouterV6BrowserTracingIntegration` within `Sentry.init` and provide the required React hooks and router functions. Then, wrap the router instance created by `createBrowserRouter` or `createMemoryRouter` with one of the following functions:
```

> To instrument your React Router, update to `Sentry.reactRouterV6BrowserTracingIntegration` within `Sentry.init` and provide the required React hooks and router functions. Then, wrap the router instance created by `createBrowserRouter` or `createMemoryRouter` with one of the following functions:

Reviewer: ```suggestion
To instrument your React Router, add `Sentry.reactRouterV6BrowserTracingIntegration()` within `Sentry.init` to your `integrations`, replacing `Sentry.browserTracingIntegration()` if it was provided, and provide the required React hooks and router functions. Then, wrap the router instance created by `createBrowserRouter` or `createMemoryRouter` with one of the following functions:
```

Reviewer: this or what @coolguyzone suggested above :)

### Usage With `<Routes />` Component

> If you're using the `<Routes />` component to define your routes, update to `Sentry.reactRouterV6BrowserTracingIntegration` inside `Sentry.init` and provide the required React hooks and router functions. Then, wrap `<Routes />` using `Sentry.withSentryReactRouterV6Routing`. This creates a higher order component, which will enable Sentry to reach your router context. You can also use `Sentry.withSentryReactRouterV6Routing` for Routes inside `BrowserRouter`, `MemoryRouter`, and `HashRouter` components.

Reviewer 2: ```suggestion
If you're using the `<Routes />` component to define your routes, update your `Sentry.browserTracingIntegration` to `Sentry.reactRouterV6BrowserTracingIntegration` inside `Sentry.init` and provide the required React hooks and router functions. Then, wrap `<Routes />` using `Sentry.withSentryReactRouterV6Routing`. This creates a higher order component, which will enable Sentry to reach your router context. You can also use `Sentry.withSentryReactRouterV6Routing` for Routes inside `BrowserRouter`, `MemoryRouter`, and `HashRouter` components.
```

);
```

This wrapping is only needed at the top level of your app, unlike React Router v4/v5, which required wrapping every `<Route />` you wanted parametrized.

### Usage With `useRoutes` Hook

_Available in `@sentry/react` version `7.12.1` and above._

> If you specify your route definitions as an object to the [`useRoutes` hook](https://reactrouter.com/en/main/hooks/use-routes), update to `Sentry.reactRouterV6BrowserTracingIntegration` inside `Sentry.init` and provide the required React hooks and router functions. Then, use `Sentry.wrapUseRoutesV6` to create a patched `useRoutes` hook that instruments your routes with Sentry.

Reviewer 2: ```suggestion
If you specify your route definitions as an object to the [`useRoutes` hook](https://reactrouter.com/en/main/hooks/use-routes), update your `Sentry.browserTracingIntegration` to `Sentry.reactRouterV6BrowserTracingIntegration` inside `Sentry.init` and provide the required React hooks and router functions. Then, use `Sentry.wrapUseRoutesV6` to create a patched `useRoutes` hook that instruments your routes with Sentry.
```

);
```

Now, Sentry should generate `pageload`/`navigation` transactions with parameterized transaction names (for example, `/teams/:teamid/user/:userid`), where applicable. This is only needed at the top level of your app, unlike React Router v4/v5, which required wrapping every `<Route />` you wanted parametrized.

### Custom Error Boundaries

When using `react-router`, errors thrown inside route elements will only be re-thrown in **development mode** while using [`strict mode`](https://react.dev/reference/react/StrictMode). In production, these errors won't be surfaced unless manually captured. If you **don't** have a custom error boundary in place, `react-router` will create a default one that "swallows" all errors.

<Alert>
  This only applies to render method and lifecycle errors since React doesn't
  need error boundaries to handle errors in event handlers.
</Alert>

> To send errors to Sentry while using a custom error boundary, use the `Sentry.captureException` method:

Reviewer: Not related to any change in this PR but generally: We should clarify here, because it is not clear to me: What If I am not using a custom error boundary, what should I do then? Should I add one to capture errors, or not? cc @lforst & @AbhiPrasad

Reviewer 3: I think for react router you need to add one, otherwise component-related errors get swallowed by the router itself.

Reviewer: yeah, so let's reword this (can be a different PR as well) to make it clear you _def_ should add this, not just if you're using a custom boundary!

Author: I  added a sentence to the first paragraph in this section (v6 and v7 pages) -- hope this is what you meant (if not, let's chat later today @mydea :) )

);
```

Now, Sentry should generate `pageload`/`navigation` transactions with parameterized transaction names (for example, `/teams/:teamid/user/:userid`), where applicable. This is only needed at the top level of your app, unlike React Router v4/v5, which required wrapping every `<Route />` you wanted parametrized.

> ## Usage With Custom Error Boundaries

Reviewer: ```suggestion
## Set Up A Custom Error Boundary
```

IMHO I would rephrase this like this, as everybody should follow this step?

Author: Updated this on v6 and v7 pages

When using `react-router`, errors thrown inside route elements will only be re-thrown in **development mode** while using [`strict mode`](https://react.dev/reference/react/StrictMode).\
In production, these errors won't surface unless captured manually. If you **don't** have a custom error boundary in place, `react-router` will create a default one that "swallows" all errors.\
Hence, to capture these errors with Sentry in production, we strongly recommend to implement a custom error boundary.

<Alert>
  This only applies to render method and lifecycle errors since React doesn't
  need error boundaries to handle errors in event handlers.

> </Alert>

Reviewer: Is this section helpful? What does it tell me? 😅 I should set this up anyhow to get all errors, I'd say?

Reviewer: I mean this whole alert here!

Author: Removed it vor v6 and v7 👍
